File: lintcode/python/0102_linked_list_cycle.py
#!/usr/bin/python -t

# linked list


"""
Definition of ListNode
class ListNode(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""

class Solution:
    """
    @param head: The first node of linked list.
    @return: True if it has a cycle, or false
    """
    def hasCycle(self, head):
        # write your code here
        if not head or not head.next:
            return False
            
        fast = head.next
        slow = head
        
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            
            if slow == fast:
                return True
                
        return False
        
        

File: lintcode/golang/0102_linked_list_cycle.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

/**
 * @param head: The first node of linked list.
 * @return: True if it has a cycle, or false
 */
func hasCycle (head *ListNode) bool {
    // write your code here
    if head == nil || head.Next == nil {
        return false
    }
    
    fast := head.Next
    slow := head
    
    for fast != nil && fast.Next != nil {
        if fast == slow {
            return true
        }
        fast = fast.Next.Next
        slow = slow.Next
    }
    
    return false
}

